### ISSUE TITLE
HttpClient - incorrect ContentType for boolean

### ISSUE THREAD

**Issue by jtdev** [angular/angular#38924]
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/common/http

### Is this a regression?

 No.

### TLDR:
A boolean as body is seen as `text/plain` where is should be seen as `application/json`, since it is valid JSON, like numbers. See:
https://jsonlint.com/
https://www.json.org/json-en.html

### Detailed description:
The `post()` method of `HttpClient` (and other request methods) accept a body method argument of type `any`. When the HTTP request is about to execute, `HttpClient` will determine the content type of `body`, so that it can set the `Content-Type` header to for example `application/json` or `text/plain`.

When an object is used as `body`, HttpClient will send it as `application/json` as expected.

When a plain number is used as `body` (e.g. `123`, without quotes), HttpClient will also send it as `application/json`, because even a plain number (without curly braces and without quotes) is valid JSON.

When a TypeScript string is used as `body` argument (e.g. `"abc"`), HttpClient will send it as `text/plain` because it will be send without quotes in the HTTP request body, and thus isn't valid JSON. So it will use `text/plain` as expected.

But when boolean is used, this behavior is different. When a plain boolean (e.g. `true`) is used as `body` argument, HttpClient will send it as `text/plain` even though a simple plain boolean is valid JSON.

The code where this problems is caused is:
https://github.com/angular/angular/blob/2686219fcee938fbb4359aa6318ffd97f91b0240/packages/common/http/src/request.ts#L295

```
    // Arrays, objects, and numbers will be encoded as JSON.
    if (typeof this.body === 'object' || typeof this.body === 'number' || Array.isArray(this.body)) {
      return 'application/json';
    }
```

Plain objects, numbers, booleans and arrays values are all [valid JSON code](https://www.json.org/json-en.html). Whereas a plain string without quotes it not valid JSON.
So plain objects, numbers, booleans and arrays values should be send as `application/json` and strings as `text/plain`. But in HttpClient this behavior unexpectedly differs for booleans.

## 🔬 Minimal Reproduction

 
https://stackblitz.com/edit/angular-ivy-umexba?file=src/app/app.component.ts

```
import { Component, VERSION } from '@angular/core';
import {HttpClient} from '@angular/common/http';

@Component({
  selector: 'my-app',
  template: `
    <button (click)="sendObject()">Send object</button>
    <button (click)="sendNumber()">Send number</button>
    <button (click)="sendBoolean()">Send boolean</button>
    <button (click)="sendString()">Send string</button>
  `
})
export class AppComponent  {
  constructor(private httpClient: HttpClient) {}

  sendObject() {
    // 'application/json'
    this.httpClient.post<any>("http://localhost", {x: 1}).subscribe();
  }

  sendNumber() {
    // 'application/json'
    this.httpClient.post<any>("http://localhost", 123).subscribe();
  }

  sendBoolean() {
    // `text/plain`, but should be 'application/json'
    this.httpClient.post<any>("http://localhost", true).subscribe();
  }

  sendString() {
    // `text/plain`
    this.httpClient.post<any>("http://localhost", "abc").subscribe();
  }
}
```

## 🌍  Your Environment

**Angular Version:**
<pre>10.1.2</pre>

**Comment by atscott** [comment]
Yep, it looks like you're right. Thanks for the detailed explanation and examples; it makes reports like this really simple to evaluate. I'm setting the priority pretty low though because the [recommendation](https://tools.ietf.org/html/rfc7493#section-4.1) is to not use texts that are neither objects nor arrays: 

> However, there are software
   implementations, coded to the older specification [RFC4627], which
   only accept JSON objects or JSON arrays at the top level of JSON
   texts.  For maximum interoperability with such implementations,
   protocol designers SHOULD NOT use top-level JSON texts that are
   neither objects nor arrays.

**Comment by gopal-jayaraman** [comment]
Looks like this issue was already fixed, is there any other reason why this is still open?
<img width="645" alt="Screen Shot 2021-04-29 at 3 59 17 PM" src="https://user-images.githubusercontent.com/78528954/116537684-28899280-a904-11eb-9550-0db6bdd86e45.png">

**Comment by atscott** [comment]
@gopal-jayaraman - This report is for `boolean`, not `number`. So the code is missing `|| typeof this.body === 'boolean'`

**Comment by gopal-jayaraman** [comment]
Thanks for the info @atscott , I will be raising a PR for this.
